日報機能・コメント機能を実装

uchihiro04 opened the pull request:
以下の機能を実装しました。

* 日報の投稿・表示・編集・削除機能
ログインユーザーが投稿していない日報は編集・削除ができない形になっています。
* 本/日報の詳細画面にそれぞれ、コメント投稿フォームと該当するページに投稿されたコメント一覧を作成
コメントがない場合は「コメントがありません」と表示されます。
* 未入力の状態で投稿ができないようにバリデーションおよび必須チェック機能を追加
* コメントの編集・削除機能を追加
ログインユーザーが投稿していないコメントは編集・削除ができない形になっています。

uchihiro04 committed:
Gemの順番を修正

rubocopの指摘を受け、Gemの順番をアルファベット順に修正しました。

uchihiro04 committed:
日報のCRUDを作成

日報を投稿・閲覧・更新・削除できるようにするために、日報のCRUDを作成しました。

uchihiro04 committed:
日報に関するページの日本語化対応を実装

日報に関連するページが英語になっているため、日本語化対応しました。

uchihiro04 committed:
メニュー欄に「日報」を追加

日報を新たに追加したため、メニュー欄に「日報」を追加しました。

uchihiro04 committed:
日報作成・更新・削除時のフラッシュメッセージを日本語化

日本語対応のため、日報作成・更新・削除時のフラッシュメッセージを日本語化しました。

uchihiro04 committed:
日報に投稿日を追加

日報に投稿日を追加しました。投稿日を日本時間に設定しました。

uchihiro04 committed:
エラー発生時の文章に日本語対応を実装

エラー発生時の文章が英文になっていたため、日本語対応を実装しました。

uchihiro04 committed:
UserモデルとReportのモデルの関連付けを追加

UserモデルとReportモデルを一対多で関連付けを追加しました。

uchihiro04 committed:
日報の更新・削除は日報を投稿した本人のみ可能な形に変更

投稿していないユーザーが日報を更新・削除できないように、日報の更新・削除は日報を投稿した本人のみ可能な形に変更しました。

uchihiro04 committed:
ページング処理を実装

日報一覧にページング処理を実装しました。

uchihiro04 committed:
不要なrespond_toを削除

jsonでアクセスを行わないため、不要なrespond_toを削除しhtmlに関するコードのみにしました。

uchihiro04 committed:
本の詳細画面にコメント表示機能・投稿機能を追加

本の詳細画面にコメント表示機能・投稿機能を追加しました。表示機能に関しては、ユーザー名とコメントが表示されます。

uchihiro04 committed:
ポリモーフィック関連付けを実装

コメント機能を実装するために、ポリモーフィック関連付けを実装しました。

uchihiro04 committed:
不要なコードを削除

ポリモーフィック関連付けの実装により不要になったコードを削除しました。

uchihiro04 committed:
コメント投稿機能・一覧表示機能の実装

本の詳細画面および日報の詳細画面にコメント投稿機能を実装しました。また、投稿されたコメントは各画面内にある「コメント一覧」に表示される形に実装しました。

uchihiro04 committed:
コメント一覧に表示する項目を変更

コメント一覧に名前（名前が未登録の場合はメールアドレス）、投稿日、コメント内容が表示されるように変更しました。

uchihiro04 committed:
コメントが未入力の状態で投稿できないように修正

コメントが未入力の状態で投稿ボタンを押すと、入力する旨を伝えるメッセージが表示されるように修正しました。

uchihiro04 committed:
N+1問題が発生するコードを修正

コメントを一覧表示する際にN+1問題が発生していたため修正しました。

uchihiro04 committed:
コメントに関する項目の日本語表示対応を実装

i18nの機能を使用して、コメントに関する項目の日本語対応を実装しました。

uchihiro04 committed:
バリデーションエラー時のレスポンスを追加

HTML5の必須チェックが効かない場合を考慮し、バリデーションエラー時のレスポンスを追加しました。

uchihiro04 committed:
不要なコメントを削除

jsonに関する不要なコメントを削除しました。

uchihiro04 committed:
コメントが1件も無い場合、コメントが無い旨を表示するように実装

コメントが1件も無い場合、「コメントがありません」と表示されるように実装しました。

uchihiro04 committed:
バリデーションを追加

NOT NULL制約がついているカラムに関して、空でないことを確認するバリデーションを追加しました。

uchihiro04 committed:
コメントの削除機能を追加

コメントの削除機能を追加しました。また、投稿者ではないコメントを削除しようとした場合、本/日報の詳細画面にリダイレクトし、フラッシュメッセージが表示されるように実装しました。

uchihiro04 committed:
コメントの一覧を部分テンプレート化

コメントの一覧について、部分テンプレート化しました。

uchihiro04 committed:
コメントの編集機能を追加

ログインしているユーザーが投稿したコメントの編集機能を追加しました。

uchihiro04 committed:
before_actionを適用するアクションを追加

before_actionにeditを追加し、Commentコントローラーのeditアクション前にメソッドが実行されるようにしました。

uchihiro04 committed:
HTML5の必須チェック機能を実装

HTML5の必須チェック機能を実装し、コメントが未入力の状態で投稿した際にメッセージが表示されるようにしました。

uchihiro04 committed:
不要になったコントローラーの共通部分を削除

不要になったコントローラーの共通部分を削除しました。

uchihiro04 committed:
rubocopからの指摘点を修正

rubocopから受けた指摘点を修正しました。

maedana commented on config/routes.rb:
```diff
@@ -2,7 +2,12 @@
+  end
+  resources :reports do
+    resources :comments, only: %i[create destroy edit update], module: 'reports'
+  end
```
books用, reports用でcontroller分けるとシンプルでいいですね :+1:

uchihiro04 replied:
ありがとうございます！

コードを書く前はルーティングも結構複雑になってしまうのかなと思っていましたが、記事や他の方の日報を参考にしながらコントローラーを分けたことで思ったよりもシンプルに書くことができました☺️

maedana commented on app/models/report.rb:
```diff
@@ -0,0 +1,7 @@
+class Report < ApplicationRecord
+  belongs_to :user
+  has_many :comments, as: :commentable, dependent: :destroy
+  validates :user_id, presence: true
```
belongs_toだけで必須チェックはかかるので上記validatesは不要ですね

uchihiro04 replied:
指摘いただいた内容を自分でも調べてみると、確かにbelongs_to関連付けを行うと必須チェックがデフォルトで設定されることが分かりました。この仕様を知りませんでした😅

[Rails5からbelongs\_toアソシエーションの挙動が変わった \- SHOYAN BLOG](https://48n.jp/blog/2017/01/13/belongs-to-should-default-to-required-true-rails5/)
[belongs\_to should default to required: true · Issue \#18233 · rails/rails](https://github.com/rails/rails/issues/18233)

コンソールでも試して必須チェックが適用されていることを確認し、バリデーションを削除いたしました。

maedana replied:
参照先にあるように仕様が変わったんですよね。昔はそれぞれ書く必要がありましたー。なお調べたということはご存知かもですが、`options: true`というのをつけると必須チェックがかからなくなります。外部キーに何らかの理由でNULLを入れたいケースで使うことがあります。

uchihiro04 replied:
外部キーの必須チェックを外すオプションに関しては、Railsガイドを読んだ際に見かけたことがありました📖　
外部キーにNULLを入れたいケースに遭遇したときはオプションのことを思い出せればと思います🙂

1点質問なのですが、`options: true`と`optional: true`はそれぞれ別物でしょうか。教えていただけますと幸いです🙇‍♂️

maedana commented on app/models/comment.rb:
```diff
@@ -0,0 +1,12 @@
+  with_options presence: true do
+    validates :content
+    validates :commentable_type
+    validates :commentable_id
```
未確認ですけど、belongs_toがあるので、commentable_typeとcommentable_idは必須チェックが自動でかかるかなーと思います

uchihiro04 replied:
commentable_typeとcommentable_idについて、バリデーションを外した後コンソールで試してみました。

```ruby
# commentable_typeとcommentable_idがnil
irb(main):003:0> comment = Comment.new(content: "hello", user_id: 1)
=> 
#<Comment:0x000000010b574570      
...     

irb(main):006:0> comment.save!
/Users/uchihiro/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/activerecord-6.1.6/lib/active_record/validations.rb:80:in `raise_validation_error': バリデーションに失敗しました: Commentableを入力してください (ActiveRecord::RecordInvalid)

# commentable_idがnil
irb(main):010:0> comment = Comment.new(content: "hello", user_id: 1, commen
table_type: 'Book')
=> 
#<Comment:0x000000010e744938                               
... 
                                                      
irb(main):012:0> comment.save!
  TRANSACTION (0.3ms)  begin transaction
  User Load (0.3ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]                                    
  TRANSACTION (0.1ms)  rollback transaction                          
/Users/uchihiro/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/activerecord-6.1.6/lib/active_record/validations.rb:80:in `raise_validation_error': バリデーションに失敗しました: Commentableを入力してください (ActiveRecord::RecordInvalid)      

# commentable_typがnil
irb(main):013:0> comment = Comment.new(content: "hello", user_id: 1, commen
table_id: 2)
=> 
#<Comment:0x000000010e90ec28                                         
...                                                                  

irb(main):015:0> comment.save!
  TRANSACTION (0.5ms)  begin transaction
  User Load (0.4ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]] 
  TRANSACTION (0.2ms)  rollback transaction
/Users/uchihiro/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/activerecord-6.1.6/lib/active_record/validations.rb:80:in `raise_validation_error': バリデーションに失敗しました: Commentableを入力してください (ActiveRecord::RecordInvalid)

# commentable_typeとcommentable_idがどちらもnilではない
irb(main):016:0> comment = Comment.new(content: "hello", user_id: 1, commen
table_type: 'Book', commentable_id: 3)
=> 
#<Comment:0x000000010e97f838         
...                                  

irb(main):018:0> comment.save!
  TRANSACTION (0.5ms)  begin transaction
  User Load (0.4ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]] 
  Book Load (1.0ms)  SELECT "books".* FROM "books" WHERE "books"."id" = ? LIMIT ?  [["id", 3], ["LIMIT", 1]] 
  Comment Create (1.6ms)  INSERT INTO "comments" ("content", "user_id", "commentable_type", "commentable_id", "created_at", "updated_at") VALUES (?, ?, ?, ?, ?, ?)  [["content", "hello"], ["user_id", 1], ["commentable_type", "Book"], ["commentable_id", 3], ["created_at", "2022-07-12 11:56:51.264163"], ["updated_at", "2022-07-12 11:56:51.264163"]]
  TRANSACTION (1.2ms)  commit transaction
=> true
```

やはりbelongs_to関連付けをしたことで、commentable_typeとcommentable_idにも必須チェックが自動でかかっているようです🤔
レビューおよびコンソールの結果を受け、バリデーションを削除いたしました。

maedana replied:
丁寧に試していて素晴らしいです :smile:

uchihiro04 replied:
ありがとうございます！　コンソールで試すことで理解が深まりました😄

maedana commented on db/migrate/20220622132113_create_reports.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :reports do |t|
+      t.string :title
+      t.text :content
```
title/contentにnot null制約がなく、model側で必須チェックもないのは意図したものでしょうか? commentsのcontentはチェックされてますし、ちょっと違和感ありますね。

uchihiro04 replied:
日報のtitleとcontentにNOT NULL制約とバリデーションをつけなかった理由といたしましては、本の登録が空白でも可能だったため、日報もタイトル・本文が未入力でも問題がないのではと実装時に考えたためです。

しかし、考え直してみるとタイトル・内容がない日報が保存できて一覧で確認できるもの不自然だと感じました。一番身近であるフィヨルドブートキャンプの日報も、タイトルと内容を空白で保存しようとするとエラーが発生することを確認しました。

考えた結果、titleとcontentにNOT NULL制約とバリデーションを追加いたしました。また、HTML5の必須チェック機能を追加いたしました。

maedana replied:
思考過程も丁寧にご返信いただけて助かります。対応内容としてもOKだと思います :+1:

uchihiro04 replied:
ありがとうございます🙇‍♂️

maedana commented on db/schema.rb:
```diff
@@ -49,6 +49,17 @@
 
+  create_table "comments", force: :cascade do |t|
+    t.text "content", null: false
+    t.integer "user_id"
```
外部キーは理由がなければNOT NULL制約かけましょうー

uchihiro04 replied:
見落としていました😵　user_idカラムにNOT NULL制約を設定しました。

maedana replied:
うっかり忘れちゃうことありますよね。なので可能な限りご自分でも`schema.rb`に変更があったときは自己レビューするようにすると抜け落ちにくくなるかもです :smile: それでも抜けるときは抜けます。僕もたまにやらかします :cry:

uchihiro04 replied:
アドバイスいただきありがとうございます！
自己レビューをすることでケアレスミスを防げるようにしていければと思います。

maedana commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,62 @@
+
+  def correct_user
+    @report = current_user.reports.find_by(id: params[:id])
+    redirect_to reports_path, alert: t('controllers.common.unauthorized_operations') if @report.nil?
```
これでもいいと思います。よくやるのは`find_by`を`find`にすることで存在しないときに`ActiveRecord::RecordNotFound`がraiseされてproductionでは404ページになるようにという対応ですねー。ご参考までに。

uchihiro04 replied:
ありがとうございます！　今回はこのままのコードで進めていきたいと思います。今後の参考にさせていただければと思います🙇‍♂️

1点確認なのですが、findメソッドを使うことによってレコードが存在しない場合にActiveRecord::RecordNotFoundが発生し、本番環境の場合`public/404.html`が表示されるという解釈で間違いないでしょうか。

一度試しに`config/environments/development.rb`内に記載された`config.consider_all_requests_local = true`をfalseに変更し、`correct_user`メソッドを以下のような形に変更したところ404ページが表示されました。
```ruby
def correct_user
  @report = current_user.reports.find(params[:id])
end
```

ご教授いただけますと幸いです🙇‍♂️

maedana replied:
そのような解釈で間違いないです!

uchihiro04 replied:
ありがとうございます！　ホッとしました☺️

uchihiro04 committed:
Reportモデルのuser_idカラムへのバリデーションを削除

belongs_to関連付けを設定した際に必須チェックがかかるため、Reportモデルのuser_idカラムへのバリデーションを削除しました。

uchihiro04 committed:
commentsテーブルのカラムのバリデーションを削除

commentsテーブルのcommentable_type、commentable_idカラムのバリデーションをそれぞれ削除しました。

uchihiro04 committed:
日報のタイトル・本文が未入力の場合投稿ができないように変更

日報のタイトル・本文が未入力の場合、「このフィールドを入力してください。」とメッセージが表示されるように変更しました。また、commentsテーブルのtitle、contentカラムにNOT NULL制約およびバリデーションを追加しました。

uchihiro04 committed:
外部キーにNOT NULL制約を追加

commentsテーブルの外部キーであるuser_idカラムにNOT NULL制約を追加しました。
